src/ast/rewriter/seq_rewriter.cpp

```diff
@@ -3344,13 +3320,137 @@ expr_ref seq_rewriter::mk_antimirov_deriv_restrict(expr* e, expr* d, expr* cond)
+// an element and its complement have the same merge-order and result in true
+expr_ref seq_rewriter::mk_regex_union_merge(expr* r1, expr* r2) {
+    expr_ref result(m());
+    expr* r1first, * r1rest, * r2first, * r2rest;
+    if (re().is_union(r1, r1first, r1rest)) {
+        if (re().is_union(r2, r2first, r2rest)) {
+            if (r1first->get_id() < r2first->get_id())
+                result = re().mk_union(r1first, mk_regex_union_merge(r1rest, r2));
+            else if (r2first->get_id() < r1first->get_id())
+                result = re().mk_union(r2first, mk_regex_union_merge(r1, r2rest));
+            else
+                result = re().mk_union(r1first, mk_regex_union_merge(r1rest, r2rest));
+        }
+        else {
+            if (r1first->get_id() < r2->get_id())
+                result = re().mk_union(r1first, mk_regex_union_merge(r1rest, r2));
+            else if (r2->get_id() < r1first->get_id())
+                result = re().mk_union(r2, r1);
+            else
+                result = r1;
+        }
+    }
+    else {
+        if (re().is_union(r2, r2first, r2rest)) {
+            if (r2first->get_id() < r1->get_id())
+                result = re().mk_union(r2first, mk_regex_union_merge(r1, r2rest));
+            else if (r1->get_id() < r2first->get_id())
+                result = re().mk_union(r1, r2);
+            else
+                result = r2;
+        }
+        else {
+            if (r1->get_id() < r2->get_id())
+                result = re().mk_union(r1, r2);
+            else if (r2->get_id() < r1->get_id())
+                result = re().mk_union(r2, r1);
+            else
+                result = r1;
+        }
+    }
+    return result;
+}
+
+expr_ref seq_rewriter::mk_regex_inter_normalize(expr* r1, expr* r2) {
+    VERIFY(m_util.is_re(r1));
+    VERIFY(m_util.is_re(r2));
+    expr_ref result(m());
+    if (r1 == r2 || re().is_empty(r1) || re().is_full_seq(r2))
+        result = r1;
+    else if (re().is_empty(r2) || re().is_full_seq(r1))
+        result = r2;
+    else if (re().is_epsilon(r1)) {
+        if (re().get_info(r2).nullable == l_true)
+            result = r1;
+        else if (re().get_info(r2).nullable == l_false)
+            result = re().mk_empty(r1->get_sort());
+        else 
+            result = mk_regex_inter_merge(r1, r2);
+    }
+    else if (re().is_dot_plus(r1) && re().get_info(r2).min_length > 0)
+        result = r2;
+    else if (re().is_dot_plus(r2) && re().get_info(r1).min_length > 0)
+        result = r1;
+    else
+        result = mk_regex_inter_merge(r1, r2);
+    return result;
+}
+// Assumes r1 and r2 are both ordered lists in right-associative form without duplicates when intersections
+// Merges the two lists as a new ordered list where duplicates are removed
+// an element and its complement have the same merge-order and result in false
+expr_ref seq_rewriter::mk_regex_inter_merge(expr* r1, expr* r2) {
```

Comment:
this code is similar to union.
Instead of duplicating code, have common routine that takes decomposition + composition as function arguments?

expr_ref seq_rewriter::merge_ac(expr* r1, expr* r2, 
std::function<bool(expr*,expr*&, expr*&)>& decompose,
std::function<expr*(expr*, expr*)>& compose) {
     ptr_vector<expr> sorted;
     expr* a, *b, *c, *d;
     while (true) {
         if (r1 == r2) {
             sorted.push_back(r1);
             break;
         }
         if (test(r1, a, b) && r2->get_id() < a->get_id()) {
             sorted.push_back(r2);
             sorted.push_back(r1);
             break;
         }
         // same for test(r2, c, d) and r1 is smaller than a
         // other cases: test(r1, a, b) and r1 == a
         // !test(r1, a, b) && !test(r2, c, d)
         if (test(r1, a, b) && test(r2, c, d) && a->get_id() < c->get_id()) {
              sorted.push_back(a);
              r1 = b;
              continue;
         
     }

     // walk sorted from back to front and apply "compose"
}

Reply:
OK, I'll rewrite the merge using a single merge routine that is non-recursive, covering both union and intersection.